crates/core/src/table/state.rs

```diff
@@ -3,23 +3,75 @@
 use std::collections::HashMap;
 use std::sync::Arc;
 
+use arrow::compute::{filter_record_batch, is_not_null};
+use arrow_array::{Array, Int64Array, StringArray, StructArray};
 use chrono::Utc;
 use futures::TryStreamExt;
 use object_store::{path::Path, ObjectStore};
 use serde::{Deserialize, Serialize};
 
 use super::config::TableConfig;
 use super::{get_partition_col_data_types, DeltaTableConfig};
+use crate::kernel::arrow::extract as ex;
 use crate::kernel::{
     Action, Add, DataType, EagerSnapshot, LogDataHandler, LogicalFile, Metadata, Protocol, Remove,
-    StructType,
+    ReplayVisitor, StructType,
 };
 use crate::logstore::LogStore;
 use crate::operations::transaction::CommitData;
 use crate::partitions::{DeltaTablePartition, PartitionFilter};
 use crate::protocol::DeltaOperation;
 use crate::{DeltaResult, DeltaTableError};
 
+pub(crate) struct AppTransactionVisitor {
+    app_transaction_version: HashMap<String, i64>,
+}
+
+impl AppTransactionVisitor {
+    pub(crate) fn new() -> Self {
+        Self {
+            app_transaction_version: HashMap::new(),
+        }
+    }
+}
+
+impl AppTransactionVisitor {
+    pub fn merge(self, map: &HashMap<String, i64>) -> HashMap<String, i64> {
+        let mut clone = map.clone();
+        for (key, value) in self.app_transaction_version {
+            clone.insert(key, value);
+        }
+
+        return clone;
+    }
+}
+
+impl ReplayVisitor for AppTransactionVisitor {
+    fn visit_batch(&mut self, batch: &arrow_array::RecordBatch) -> DeltaResult<()> {
+        if batch.column_by_name("txn").is_none() {
+            return Ok(());
+        }
+
+        let txn_col = ex::extract_and_cast::<StructArray>(batch, "txn")?;
+        let filter = is_not_null(txn_col)?;
+
+        let filtered = filter_record_batch(batch, &filter)?;
+        let arr = ex::extract_and_cast::<StructArray>(&filtered, "txn")?;
+
+        let id = ex::extract_and_cast::<StringArray>(arr, "appId")?;
+        let version = ex::extract_and_cast::<Int64Array>(arr, "version")?;
+
+        for idx in 0..id.len() {
+            let app = ex::read_str(id, idx)?;
+            let version = ex::read_primitive(version, idx)?;
+
+            self.app_transaction_version.insert(app.to_owned(), version);
+        }
+
+        Ok(())
+    }
```

Comment:
Never actually worked with the transaction IDs, but I think we may have to insert transactrion Ids only when its not yet inserted. Reason being that we replay the log from highes commit to lowest.